### DISCUSSION TITLE
[Item] collect annotations and notes by tag

### DISCUSSION THREAD

**Discussion by windingwind** [windingwind/zotero-better-notes#236] [Note Templates]
Author: @windingwind  
Discuss: https://forums.zotero.org/discussion/comment/414541/#Comment_414541  
> Input tags once and search annotations & notes with these tags in the selected items.

- Item 1
  - Tag 1
    - note link 1
    - annotation 1
  - Tag 2
     - annotation 2
   - ...
- Item 2
  - Tag 1
    - note link 2
  - Tag 2
    - annotation 3

![image](https://user-images.githubusercontent.com/33902321/184292137-c5d44fdf-cb0a-4fe5-af75-2fe63758f29d.png)

```yaml
# This template is specifically for importing/sharing, using better 
# notes 'import from clipboard': copy the content and
# goto Zotero menu bar, click Edit->New Template from Clipboard.  
# Do not copy-paste this to better notes template editor directly.
name: "[Item] test"
content: |-
  // @author windingwind
  // @link https://github.com/windingwind/zotero-better-notes/discussions/236
  // @beforeloop-begin
  ${{
  	const tagRaw = Zotero.getMainWindow().prompt("Please input tags. Split with ',':", "");
  	sharedObj.tags = tagRaw.split(',').filter(t => t.length);
  	return "";
  }}$
  // @beforeloop-end
  // @default-begin
  ${{
  	async function getAnnotationsByTag(item, tag) {
  		let annots = item.getAnnotations();
  		annots = tag.length ?
  			annots.filter((annot) => annot.getTags().map(tagObj => tagObj.tag).includes(tag)) :
  			annots;
  		return await Zotero.BetterNotes.api.convert.annotations2html(annots, { noteItem: targetNoteItem });
  	}
  	if (!sharedObj.tags.length) {
  		return "";
  	}
  	const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
  		i.isPDFAttachment()
  	);
  	let res = "";
  	res += `<h1>${topItem.getField('title')}</h1>`;
  	for (let attachment of attachments) {
  		res += `<h2>${attachment.getFilename()}</h2>`;
  		for (const tag of sharedObj.tags) {
  			res += `<h3>Tag: ${tag}</h3>`;
  			const annots = (await getAnnotationsByTag(attachment, tag));
  			res += annots ? annots : "<p>No result</p>";
  		}
  	}
  	res += `<h2>Notes</h2>`;
  	const notes = itemNotes.filter(noteItem => {
  		for (const tag of sharedObj.tags) {
  			if (noteItem.getTags().map(tagObj => tagObj.tag).includes(tag)) {
  				return true;
  			}
  		}
  		return false;
  	});
  	const notesWithTags = `${notes.map((noteItem) => {
  		const noteLink = Zotero.BetterNotes.api.convert.note2link(noteItem);
  		const noteLine = `<p style="color:red; background-color: #efe3da;">📜 Article Note:  <a href="${
  				noteLink
  			}">${noteItem.getNoteTitle() || noteLink}</a></p>
  <p>tags: ${noteItem.getTags().map(tagObj => tagObj.tag).join(', ')}</p>
  <p> </p>`;
  		return noteLine;
  	}).join("\n")}`
  	res += notesWithTags || '<p>No result</p>'
  	return res;
  }}$
  // @default-end
```

**Comment by cAlting** [comment]
Hi,

I always get the error "ReferenceError: prompt is not defined". I guess that I'm missing something, because all the templates that contain a prompt give me this error and all templates containing a "if (!editor)" give me another error (something like editor is not defined).

**Reply by windingwind** [reply]
The template is updated. Pass test on BN 1.1.4-12.

**Reply by yzy-lex** [reply]
Thank you very much! But can this template only output items that have result?

**Comment by mjthoraval** [comment]
Can this template (or an older version of it) still work with Zotero 6?

This Note Template has the tag `support BN 1.0`.
It suggests the template should work with BN 1.0: [Release 1.0.4](https://github.com/windingwind/zotero-better-notes/releases/tag/1.0.4), corresponding to Zotero 6 (Zotero 6.0.30).

But when I try this template in Zotero 6.0.30, I get the following error:
```
[JavaScript Application]
Template [Item] test Error: SyntaxError: missing : after property id
```
![image](https://github.com/windingwind/zotero-better-notes/assets/36054098/f8094fda-4e87-4a79-abb9-d829e4f0b12c)

I was able to run it in Zotero 7, so the problem seems to be limited to Zotero 6.

Does the tag refer to some other functionality that was added in BN 1.0?
If not compatible with Zotero 6, maybe the tag should be updated to `Support BN 1.1+` or `Zotero 7 only`? Or only tag the ones that are `Zotero 6 compatible` if the majority of the templates are for Zoteor 7?

Are there still old Note Templates that are still compatible with Zotero 6?

**Reply by windingwind** [reply]
Yes, since these templates use the new feature `${{}}$` for multi-line scripts, it's not compatible with Z6 anymore. When I assigned them with that label, I didn't expect Z7 would still be not ready after a year.

**Reply by mjthoraval** [reply]
Thank you for your reply.
This template was posted before Z7: [`May 21, 2023`](https://forums.zotero.org/discussion/105094/announcing-the-zotero-7-beta/p1)

So I was expecting that an old version of the script would still be compatible with Z6.
I have tried to see in the history of the post to get the original version. But that also did not work.
![image](https://github.com/windingwind/zotero-better-notes/assets/36054098/3986db59-816d-4582-82f2-b554172a33d3)

Am I copying the wrong template?
Is there an older version of this template that should work on Z6?

**Reply by windingwind** [reply]
That's for BN <0.8.9 I think.

You can replace `prompt(...)` with `window.prompt(...)` in the template code to fix it.

**Reply by mjthoraval** [reply]
Thank you.

**Reply by dexter-gh** [reply]
The quick fix mentioned does not fix the error for me. To ask more specifically, is your update related to the following line of code:

`const tagRaw = Zotero.getMainWindow().prompt("Please input tags. Split with ',':", "");`

**Comment by SimonGideonRay** [comment]
Hello, the code does not work in Zotero 7.0.0-65 and bn1.1.4-57. The feature you developed is very useful. Could you please guide me on how to modify it to make it work?

I added the code you provided into the bn template, and then created the note in an entry. However, the note added are actually empty, and it does not generate notes based on tags.

look forward to your assistance and reply! thanks!

**Reply by windingwind** [reply]
I confirm it works. You'll need to make sure 
1. the template name starts with `[item]`;
2. there are tagged annotations under the selected item's attachment;
3. the tags you input belong to at least one of the annotations

otherwise no annotation will be found and the note will be left empty

**Reply by MaxVonMustermann** [reply]
@windingwind 
I am responding to this thread based on another that has been closed: https://github.com/windingwind/zotero-better-notes/issues/1092
Creating a note based on the template works npw. 
On question: I would like to import all annotations, no matter what tag they belong to. Is this somehow possible? Basically having a button to "import all existing" tags in this project + untagged items. This would be really helpful!
I am looking forward to your response, thank you very much!

**Reply by windingwind** [reply]
> @windingwind I am responding to this thread based on another that has been closed: #1092 Creating a note based on the template works npw. On question: I would like to import all annotations, no matter what tag they belong to. Is this somehow possible? Basically having a button to "import all existing" tags in this project + untagged items. This would be really helpful! I am looking forward to your response, thank you very much!

Sorry, sometimes I just miss the email...

Yes it's possible. Simply remove the following lines:

```
annots = tag.length ?
  			annots.filter((annot) => annot.getTags().map(tagObj => tagObj.tag).includes(tag)) :
  			annots;
```

**Comment by sjgknight** [comment]
Is it possible to modify this template so it imports the full content of the notes? I had an earlier template that did this but I'm not sure what needs changing to make it compatible with zotero 7/more recent bn. Any help much appreciated!

**Reply by windingwind** [reply]
Yes. Simply replace the `note2link` with `note2html` (with the target note passed properly). Let me know if you need further help.

**Comment by Ravenclawer** [comment]
Can I import annotations that are grouped under a “tag” heading1 —its items and notes formatted as subheadings—while hiding any unmatched items?

**Reply by Ravenclawer** [reply]
and it's better to support multiple tags and exclusion of tags

**Reply by windingwind** [reply]
though i don't exactly understand what you want, i assume LLMs would be helpful here
